pfcpiface/parser.go

```diff
@@ -0,0 +1,257 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright(c) 2020 Intel Corporation
+
+package main
+
+import (
+	"log"
+	"net"
+
+	"github.com/wmnsk/go-pfcp/ie"
+	"github.com/wmnsk/go-pfcp/message"
+)
+
+func parseCreatePDRPDI(pdi []*ie.IE) (srcIface uint8, teid uint32, ueIP4 net.IP) {
+	var err error
+
+	for _, ie2 := range pdi {
+		switch ie2.Type {
+		case ie.SourceInterface:
+			srcIface, err = ie2.SourceInterface()
+			if err != nil {
+				log.Println("Failed to parse Source Interface IE!")
+			} else {
+				if srcIface == ie.SrcInterfaceCPFunction {
+					log.Println("Detected src interface cp function. Ignoring for the time being")
+				}
+			}
+		case ie.FTEID:
+			fteid, err := ie2.FTEID()
+			if err != nil {
+				log.Println("Failed to parse FTEID IE")
+			} else {
+				teid = fteid.TEID
+			}
+		case ie.UEIPAddress:
+			ueip4, err := ie2.UEIPAddress()
+			if err != nil {
+				log.Println("Failed to parse UE IP address")
+			} else {
+				ueIP4 = ueip4.IPv4Address
+			}
+		case ie.SDFFilter:
+			// Do nothing for the time being
+		case ie.QFI:
+			// Do nothing for the time being
+		}
+	}
+	return srcIface, teid, ueIP4
+}
+
+func parseCreatePDR(ie1 *ie.IE, fseid *ie.FSEIDFields) (pdrI *pdr) {
+	var srcIface uint8
+	var teid uint32
+	var ueIP4 net.IP
+
+	/* reset outerHeaderRemoval to begin with */
+	outerHeaderRemoval := uint8(0)
+
+	pdrID, err := ie1.PDRID()
+	if err != nil {
+		log.Println("Could not read PDR!")
+		return nil
+	}
+
+	pdi, err := ie1.PDI()
+	if err != nil {
+		log.Println("Could not read PDI!")
+		return nil
+	}
+
+	_, err = ie1.OuterHeaderRemoval()
+	if err == nil {
+		res, err := ie1.OuterHeaderRemovalDescription()
+		if res == 0 && err == nil { // 0 == GTP-U/UDP/IPv4
+			outerHeaderRemoval = 1
+		}
+	}
+
+	// parse PDI IE and fetch srcIface, teid, and ueIPv4Address
+	srcIface, teid, ueIP4 = parseCreatePDRPDI(pdi)
+
+	farID, err := ie1.FARID()
+	if err != nil {
+		log.Println("Could not read FAR ID!")
+		return nil
+	}
+
+	// uplink PDR may not have UE IP address IE
+	// FIXIT/TODO/XXX Move this inside SrcInterfaceAccess IE check??
+	var ueIP uint32
+	var ueIPMask uint32
+	if len(ueIP4) == 0 {
+		ueIP = 0
+		ueIPMask = 0
+	} else {
+		ueIP = ip2int(ueIP4)
+		ueIPMask = 0xFFFFFFFF
+	}
+
+	// populated everything for PDR, and set the right pdr_
+	if srcIface == ie.SrcInterfaceAccess {
+		pdrU := pdr{
+			srcIface:     access,
+			eNBTeid:      teid,
+			dstIP:        ueIP,
+			srcIfaceMask: 0xFF,
+			eNBTeidMask:  0xFFFFFFFF,
+			dstIPMask:    ueIPMask,
```

Comment:
pull this inside the `parseCreatePDRPDI` and make it return a pdr

Reply:
Looks  better in current form.